sw/host/opentitanlib/src/bootstrap/legacy.rs

```diff
@@ -202,40 +206,74 @@ impl UpdateProtocol for Legacy {
     fn update(&self, spi: &dyn Target, payload: &[u8]) -> Result<()> {
         let frames = Frame::from_payload(payload);
 
-        for (i, frame) in frames.iter().enumerate() {
-            let want_hash = frames[i.saturating_sub(1)].frame_hash();
-            // Repeatedly transmit the frame, until receiving expected response.
-            loop {
-                // TODO: Introduce a progress callback, to allow common code to show progress bar.
-                eprint!("{}.", i);
-
-                // Important? Comment in spiflash.cc indicates that it seems to corrupt without a
-                // 1ms delay.
-                std::thread::sleep(self.inter_frame_delay);
-
-                // Write the frame and read back the hash of the previous frame.
-                let mut response = [0u8; std::mem::size_of::<Frame>()];
-                spi.run_transaction(&mut [Transfer::Both(frame.as_bytes(), &mut response)])?;
-
-                if response.iter().all(|&x| x == response[0]) {
-                    // A response consisteing of all identical bytes is a status code.
-                    match LegacyBootstrapError::from(response[0]) {
-                        LegacyBootstrapError::NotReady => continue, // Retry sending same frame.
-                        error => return Err(error.into()),
+        // All frames up to but not including this index have been ack'ed by the bootloader.
+        // Once this reaches frames.len(), the operation was successful.
+        let mut first_unacked_index = 0;
+
+        // Counts the number of repeated failures in getting a frame ack.
+        let mut consecutive_errors = 0;
+
+        // Set if the past transaction ack'ed the block sent before that one, while transmitting
+        // the next.  If so, we heuristically assume that the block just sent will be ack'ed in
+        // the upcoming bidirectional transaction.
+        let mut optimistic = false;
+        
+        // Set if the past transaction ack'ed the block seet before that one, while
```

Comment:
s/seet/seen/ or s/seet/sent/ ?

Reply:
I have elaborated a little on this comment.